EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -0,0 +1,375 @@
+                if (psm != null)
+                {
+                    List<int> indiciesOfPeptidesToRemove = new List<int>();
+                    List<(int notch, PeptideWithSetModifications pwsm)> bestMatchingPeptidesToRemove = new List<(int notch, PeptideWithSetModifications pwsm)>();
+                    List<double> pepValuePredictions = new List<double>();
+
+                    //Here we compute the pepvalue predection for each ambiguous peptide in a PSM. Ambiguous peptides with lower pepvalue predictions are removed from the PSM.
+                    foreach (var (Notch, Peptide) in psm.BestMatchingPeptides)
+                    {
+                        PsmData pd = CreateOnePsmDataFromPsm2(psm, Notch, Peptide, accessionAppearances, sequenceToPsmCount);
+
+                        var pepValuePrediction = predictionEngine.Predict(pd);
+
+                        someOut.Add(pd.AccessionAppearances.ToString() + "|" + pd.Ambiguity.ToString() + "|" + pd.DeltaScore.ToString() + "|" + pd.Intensity.ToString() + "|" + pd.Label + "|" + pd.LongestFragmentIonSeries + "|" + pd.MissedCleavagesCount + "|" + pd.ModsCount + "|" + pd.Notch + "|" + pd.PsmCount + "|" + pd.ScanPrecursorCharge + "|" + pepValuePrediction.Prediction + "|" + pepValuePrediction.Probability + "|" + pepValuePrediction.Score);
+
+                        pepValuePredictions.Add(pepValuePrediction.Probability);
+                        //A score is available using the variable pepvaluePrediction.Score
+                    }
+
+                    double highestPredictedPEPValue = pepValuePredictions.Max();
+                    int numberOfPredictions = pepValuePredictions.Count - 1;
+
+                    for (int i = numberOfPredictions; i >= 0; i--)
+                    {
+                        if (Math.Abs(highestPredictedPEPValue - pepValuePredictions[i]) > 0.000001)
+                        {
+                            indiciesOfPeptidesToRemove.Add(i);
+                            pepValuePredictions.RemoveAt(i);
+                            //pValuePredictionStrings.RemoveAt(i);
+                        }
+                    }
+
+                    int index = 0;
+
+                    foreach (var (Notch, Peptide) in psm.BestMatchingPeptides)
+                    {
+                        if (indiciesOfPeptidesToRemove.Contains(index))
+                        {
+                            bestMatchingPeptidesToRemove.Add((Notch, Peptide));
+                        }
+                        index++;
+                    }
+
+                    foreach (var (notch, pwsm) in bestMatchingPeptidesToRemove)
+                    {
+                        ambiguousScans = ambiguousScans + psm.ScanNumber + "|";
+                        psm.RemoveThisAmbiguousePeptide(notch, pwsm);
+                    }
+
+                    psm.FdrInfo.PEP = pepValuePredictions[0]; //they should all be the same at this point so it doesn't matter which you take. First is good.
```

Comment:
I think you want
`psm.FdrInfo.PEP = 1 - pepValuePredictions[0];`

the PEP is the probability of being an error. the machine learning algorithm is predicting the probability of being correct, I think?